File: backtracking/0077-combinations/solution.java
// 77. Combinations
// https://leetcode.com/problems/combinations/
// Medium | Java | Accepted 2026-08-22
// Runtime 14 ms | Memory 97.9 MB

class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    int kk;
    int nn;
    public List<List<Integer>> combine(int n, int k) {
        nn = n;
        kk = k;
        List<Integer> temp = new ArrayList<>();
        backtrack(temp, 1);
        return ans;
    }

    public void backtrack(List<Integer> temp, int start)
    {
        if(temp.size()==kk)
        {
            ans.add(new ArrayList<>(temp));
            return;
        }
        for(int i = start; i<=nn-(kk-temp.size())+1; i++)
        {
            temp.add(i);
            backtrack(temp, i+1);
            temp.remove(temp.size()-1);
        }
    }
}

File: backtracking/0077-combinations/solution.cpp
// 77. Combinations
// https://leetcode.com/problems/combinations/
// Medium | C++ | Accepted 2026-07-16
// Runtime 74 ms | Memory 82.5 MB

class Solution {
public:
    vector<vector<int>> res;
    int nn;
    int kk;
    vector<vector<int>> combine(int n, int k) {
        nn = n;
        kk = k;
        vector<int> vec;
        backtrack(vec, 1);
        return res;
    }

    void backtrack(vector<int>& temp, int i)
    {
        if(temp.size()==kk)
        {
            res.push_back(temp);
            return;
        }
        for(int j = i; j<=nn; j++)
        {
            temp.push_back(j);
            backtrack(temp, j+1);
            temp.pop_back();
        }
    }
};
